File: gpgga2txt.jl
# gpgga2txt.jl

# Extracts date, latitude, and longitude, from NMEA formatted GPS text files. Saves the output to a
# text file in the local directory.

# Note:
# -Accurate datetime extraction assumes that the ship is using RVDAS for writing the timestamp. This
# code has been tested on several UNOLS ships successfully, but datestamp format has been known to 
# change on non UNOLS research vessels. 

# Input			: Filepath

# Output 		: gpgga.txt with the following fields
#				-Datetime (UTC): yyyy,mm,dd,HH,MM,SS.SS
#				-Longitude (Decimal)
#				-Latitude (Decimal)

# Created on: 10/20/2023
# Last edit: 11/01/2023
# Michael Cappola (mcappola@udel.edu)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

using Dates

function gpgga2txt(filename)

open("gpgga.txt","w") do file

	# Write Headerline
	println(file,"Year,Month,Day,Hour,Minute,Second,Latitude,Longitude")

	# Loop through data file
	for line in eachline(filename)
		
		# If GPGGA line, extract GPS information
		if occursin("GPGGA",line)

			# Split GPGGA line by comma into the elements array
			elements = split(line,",")
			
			# Checks if GPS was on
			if !isempty(elements[3]) || !isempty(elements[5])

				# Extract date elements	
				datestring = elements[1]
				datestring = split(datestring," ")
				datestring = datestring[1]
				dateelem = split(datestring,":")
			
				# Make year from yy+ddd			
				yearday = split(dateelem[1],"+")
				year = 2000 + parse(Int16,yearday[1])
			
				# Make month and day from yy+ddd
				t = Dates.value(Date(year,1,1))
				t = t + parse(Int16,yearday[2]) - 1
				month = Dates.month(t)
				day = Dates.day(t)			

				# Extract hour, min, and sec			
				hour = parse(Int16,dateelem[2])
				mmm = parse(Int16,dateelem[3])
				sec = parse(Float16,dateelem[4])
				sec = round(sec, digits = 2)

				# Extract Latitude and convert to decimal			
				lat = elements[3]
				lat = parse(Float64,lat)
				deg = floor(lat/100)
				min = ((lat/100)-deg)*100
				dec = min/60
				com = deg + dec			
				if elements[4] == "N"
					latitude = com
				else
					latitude = com*-1
				end

				# Extract Longitude and convert to decimal
				lon = elements[5]
				lon = parse(Float64,lon)
				deg = floor(lon/100)
				min = ((lon/100)-deg)*100
				dec = min/60
				com = deg + dec
				if elements[6] == "E"
					longitude = com
				else
					longitude = com*-1
				end
				
				latitude = round(latitude, digits = 4)
				longitude = round(longitude, digits = 4)	

			else
				year = NaN
				month = NaN
				day = NaN
				hour = NaN
				min = NaN
				sec = NaN
				latitude = NaN
				longitude = NaN
			end
			
			# Assemble datastring
			datastring = string(year) * "," * string(month)
			datastring = datastring * "," * string(day)
			datastring = datastring * "," * string(hour)
			datastring = datastring * "," * string(mmm)
			datastring = datastring * "," * string(sec)
			datastring = datastring * "," * string(latitude)
			datastring = datastring * "," * string(longitude)
			
			# Write data to file
			println(file,datastring)				
		end
	end
end

end

for arg in ARGS
	filename = arg
	gpgga2txt(filename)
end

File: gpgga2txt.m
function gpgga2txt(file,varargin)

% Extracts date, latitude, and longitude, from NMEA formatted GPS text
% files (Text files with a NMEA GPGGA line). Saves output to a text file.

% Notes
% - Accurate datetime extraction assumes the ship is using RVDAS for
% writing the timestamp. This code has been tested on several UNOLS ships
% successfully. 

% Required Argument     : Filepath

% Optional Arguments    : OutputFilename - Default is gppga.txt
%                       : Truncate - Number of skipped data lines.
%                       Decreases data output density. Default is 1 which 
%                       keeps every scan. 10 would save every 10th scan.

% Output                : DateTime (UTC)
%                       : Longitude (Decimal)
%                       : Latitude (Decimal)

% Written by: Michael Cappola (mcappola@udel.edu)
% Created on: 03/27/2022
% Last edit: 10/28/2023
% ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

p = inputParser;
addRequired(p,'file');
addParameter(p,'OutputFilename','gpgga.txt');         
addParameter(p,'Truncate',1);  
parse(p,file,varargin{:});

filename = p.Results.OutputFilename;
truncate = p.Results.Truncate;

% Counts the number of lines containing "GPGGA" for preallocating arrays.
% Saves time for big files.
fid = fopen(file);
tline = fgetl(fid);
n = 0;
while ischar(tline)
    if contains(tline,'GPGGA')      
        n=n+1;     
    end
    tline = fgetl(fid);
end
fclose(fid);

% Format of "GPGGA" Line.
fmt = '%f+%f:%f:%f:%f %c%c%c%c%c%c,%f,%f,%c,%f,%c,%f,%f,%f,%f,%c,%f,%c,%s';

% Preallocation.
dat = nan(floor(n/truncate),6);
lat = nan(floor(n/truncate),1);           
lon = nan(floor(n/truncate),1);

% Pulls out GPS Data.
fid = fopen(file);
tline = fgetl(fid);
n = 0;
k = 0;
while ischar(tline)
    if contains(tline,'GPGGA')
        n=n+1;
        if rem(n,truncate)==0
            k = k+1;
            a = textscan(tline,fmt);    % Builds char of selected line w/ fmt.
            
            %Date
            yr = a{1} + 2000;           % Convert RVDAS Year fmt to year
            dt = datevec(datenum(yr, ones(size(yr)), a{2}));    %ddd -> mmdd
            mm = dt(:,2);               % Mon
            dd = dt(:,3);               % Day
            dt(4) = a{3};               % Hr
            dt(5) = a{4};               % Min
            dt(6) = a{5};               % Sec
            dat(k,:) = dt;
            
            if ~isempty(a{13}|a{15})            % Checks if GPS is running.
                %Latitude
                deg = floor(a{13}/100);         % Pull Degree out
                min = (a{13}/100 - deg)*100;    % Pull Minute out
                dec = min/60;                   % Convert Minute to Dec
                com = deg + dec;                % Combine
                if a{14} == 'N'                 % N = +, S = -.
                    lat(k,:) = com;
                else
                    lat(k,:) = com*-1;
                end
                
                %Longitude
                deg = floor(a{15}/100);         % Pull Degree out
                min = (a{15}/100 - deg)*100;    % Pull Minute out
                dec = min/60;                   % Convert Minute to Dec
                com = deg + dec;                % Combine
                if a{16} == 'E'                 % E = +, W = -.
                    lon(k,:) = com;
                else
                    lon(k,:) = com*-1;
                end
            else
                lat(k,:) = nan;                 % Records nan if GPS is off.
                lon(k,:) = nan;
            end            
        end
    end
    tline = fgetl(fid);
end
fclose(fid);

dat = datetime(dat);

% Sort by date, just incase the files we appended out of order.
[dat,idx] = sort(dat);
lat = lat(idx);
lon = lon(idx);

% Clean up variables and filenames for text file.
dat = datevec(dat);
Year = dat(:,1);
Month = dat(:,2);
Day = dat(:,3);
Hour = dat(:,4);
Minute = dat(:,5);
Second = dat(:,6);

Latitude = round(lat,4);
Longitude = round(lon,4);

% Makes and saves table for text file.
t = table(Year,Month,Day,Hour,Minute,Second,Latitude,Longitude);
writetable(t,filename);

